### PR TITLE
azure-cli: fix transitive dependencies

### LINKED ISSUE #281545: Build failure: azure-cli
### Steps To Reproduce

1. build *azure-cli* against unstable or master
`nix build --expr '(import (builtins.getFlake "github:nixos/nixpkgs/e862ac6b5ca280d3e41db71467f129839e47ee45") { system = "x86_64-linux"; }).azure-cli'`

### Build log

https://gist.github.com/CryoMyst/63c121d42ac865809e950d90c1410384

### Notify maintainers

@AkechiShiro @jonringer

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.8, NixOS, 24.05 (Uakari), 24.05.20240113.9b19f5e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by katexochen**
## Description of changes

Seems like these are broken after
https://github.com/NixOS/nixpkgs/pull/279138
https://github.com/NixOS/nixpkgs/pull/279143

(somewhat suspicious that it built on the PR?)

---

There is also an outstanding update in https://github.com/NixOS/nixpkgs/pull/281183 with some discussion.

Cherry-picked the fix of the build failure into this PR as the update isn't going that smooth.

fixes https://github.com/NixOS/nixpkgs/issues/281545

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by katexochen** [comment]
Result of `nixpkgs-review pr 281213` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>azure-cli</li>
    <li>azure-cli.dist</li>
  </ul>
</details>

**Comment by katexochen** [comment]
> Seems like these are broken after
> https://github.com/NixOS/nixpkgs/pull/279138
> https://github.com/NixOS/nixpkgs/pull/279143
> 
> (somewhat suspicious that it built on the PR?)

Actually, it still builds fine on master after these two were merged. Something else must have happened. I tried to bisect, but there was a staging merge in between and most of the commit are broken in other ways...

**Comment by AkechiShiro** [comment]
@katexochen
I believe this commit might be the source of the build failure : https://github.com/NixOS/nixpkgs/commit/9490738be7b6a33844e4faf192f168bb755539e6 (source : https://github.com/NixOS/nixpkgs/issues/281164#issuecomment-1892754362)

**Comment by katexochen** [comment]
> I believe this commit might be the source of the build failure : [9490738](https://github.com/NixOS/nixpkgs/commit/9490738be7b6a33844e4faf192f168bb755539e6) (source : [#281164 (comment)](https://github.com/NixOS/nixpkgs/issues/281164#issuecomment-1892754362))

@AkechiShiro Reverting the commit on master doesn't seem to fix the build failure.

**Comment by katexochen** [comment]
@AkechiShiro the error I'm seeing is
```
       > Checking runtime dependencies for azure_mgmt_batch-17.0.0-py3-none-any.whl
       >   - msrest not installed
```
which is the same as on master.

**Comment by gemone** [comment]
same issue

**Comment by AkechiShiro** [comment]
It seems that the issue comes from this breaking change that took place in  https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/40?page=2, most likely due to this commit : https://github.com/NixOS/nixpkgs/commit/8f3162f83fcccf26bf50c55ff960fd2dee5264b2

**Comment by katexochen** [comment]
> It seems that the issue comes from this breaking change that took place in https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/40?page=2, most likely due to this commit : [8f3162f](https://github.com/NixOS/nixpkgs/commit/8f3162f83fcccf26bf50c55ff960fd2dee5264b2)

Ahh! That seem to make sense. :) 

@AkechiShiro can you review, so we can merge the fix to get it building on main again?

**Comment by AkechiShiro** [comment]
Sure, I will review the merge in a few hours (at least by tonight)

**Review by msanft — APPROVED** [review]
Builds and launches properly on ` x86_64-linux`

File: pkgs/tools/admin/azure-cli/python-packages.nix [hunk 1]
```diff
@@ -139,6 +137,11 @@ let
       azure-mgmt-appcontainers = overrideAzureMgmtPackage super.azure-mgmt-appcontainers "2.0.0" "zip"
         "sha256-ccdIdvdgTYPWEZCWqkLc8lEuMuAEERvl5B1huJyBkvU=";
 
+      azure-mgmt-batch = (overrideAzureMgmtPackage super.azure-mgmt-batch "17.0.0" "zip"
+        "sha256-hkM4WVLuwxj4qgXsY8Ya7zu7/v37gKdP0Xbf2EqrsWo=").overridePythonAttrs (attrs: {
+        propagatedBuildInputs = attrs.propagatedBuildInputs or [ ] ++ [ self.msrest self.msrestazure ];
```
Comment (AkechiShiro) [inline pkgs/tools/admin/azure-cli/python-packages.nix:142]: `msrestazure` is necessary ?
Also why is it outside of the package derivation ? Shouldn't it be inside the package derivation as a dependency ?

**Comment by AkechiShiro** [comment]
I did test the PR it works all fine, but I still would like to know why are needed : 
- msrestazure
- and azure-mgmt-nspkg

I believe removing both do not break the build process/nor self-test.

**Comment by AkechiShiro** [comment]
I did more testing.

- `msrestazure` is indeed needed or it will cause failure during the build :
```
error: builder for '/nix/store/v5qq4qbqislfl7d05gzkrdjvnj2saqbp-python3.11-azure-mgmt-kusto-0.3.0.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
       >   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
       >   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
       >   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
       >   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
       >   File "/nix/store/b9srnzghb7pkyhwdl63azzynnc40gqhk-python3.11-azure-mgmt-kusto-0.3.0/lib/python3.11/site-packages/azure/mgmt/kusto/__init__.py", line 12, in <module>
       >     from .kusto_management_client import KustoManagementClient
       >   File "/nix/store/b9srnzghb7pkyhwdl63azzynnc40gqhk-python3.11-azure-mgmt-kusto-0.3.0/lib/python3.11/site-packages/azure/mgmt/kusto/kusto_management_client.py", line 14, in <module>
       >     from msrestazure import AzureConfiguration
       > ModuleNotFoundError: No module named 'msrestazure'
       For full logs, run 'nix log /nix/store/v5qq4qbqislfl7d05gzkrdjvnj2saqbp-python3.11-azure-mgmt-kusto-0.3.0.drv'.
error (ignored): error: cannot unlink '/tmp/nix-build-python3.11-azure-mgmt-recoveryservicesbackup-7.0.0.drv-0': Directory not empty
error: 1 dependencies of derivation '/nix/store/y62542bk42a9pkbdc30wxp66mm2wdiql-python3.11-azure-cli-2.55.0.drv' failed to build
```

- `azure-mgmt-nspkg` seems to not be needed, build and self-test of the CLI passes fine. So I wonder why it was put in the first place hmm.

**Comment by katexochen** [comment]
My original intention was just to quickly restore what worked before.

But I did some research and the requirements are now reduced to the minimum, based on

setup.py of azure-mgmt-servicefabric v1.0.0:
https://github.com/Azure/azure-sdk-for-python/blob/19e35d6472e40ca8974420a4024d7cf34160e25a/sdk/servicefabric/azure-mgmt-servicefabric/setup.py#L80-L82

setup.py of azure-mgmt-imagebuilder v1.2.0:
https://github.com/Azure/azure-sdk-for-python/blob/8a8673b5fdb3952cc3f50be2e500a1b77b72f120/sdk/compute/azure-mgmt-imagebuilder/setup.py#L73-L76C55

setup.py of azure-mgmt-batch v17.0.0:
https://github.com/Azure/azure-sdk-for-python/blob/9f156e77f161b7f95d1e7c78a8359f6a994046b5/sdk/batch/azure-mgmt-batch/setup.py#L72-L75

**Comment by AkechiShiro** [comment]
Have you tested building @katexochen ? 
You have not run into this error with `msrestazure`  https://github.com/NixOS/nixpkgs/pull/281213#issuecomment-1897064404 ?

**Comment by katexochen** [comment]
> Have you tested building @katexochen ? You have not run into this error with `msrestazure` [#281213 (comment)](https://github.com/NixOS/nixpkgs/pull/281213#issuecomment-1897064404) ?
> 
> Will do thorough check by tonight/review, hopefully we'll merge tonight as well, so master gets the fix.

Yes, builds fine locally and seems to also build in CI. The error you posted seems to be about `python3.11-azure-mgmt-kusto` which wasn't touched in this PR.

**Review by jonringer — APPROVED** [review]
was able to get usage.

Thanks for fixing this up :)

```
https://github.com/NixOS/nixpkgs/pull/281213

2 packages built:
azure-cli azure-cli.dist
```

**Comment by AkechiShiro** [comment]
Just want to let you know @katexochen that 2.56.0 tests are still failing if we take up this PR as the base and update.

I will try and investigate further by pinning all the `requirements.txt` (tag for the 2.56.0 release upstream) dependencies this week-end hopefully.

**Pull request merged**